tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lower_alloc_dealloc.cc

```diff
@@ -0,0 +1,203 @@
+
+namespace mlir {
+namespace disc_ral {
+
+using LLVM::LLVMFuncOp;
+
+static constexpr const char* kMalloc = "alloc";
+static constexpr const char* kFree = "free";
+
+// A rewrite pattern to convert memref.alloc operations into corresponding
+// runtime wrapper calls (modeled by ral.dispatch ops)
+class ConvertMemRefAllocOpToDispatchOpPattern
+    : public ConvertOpToLLVMPattern<memref::AllocOp> {
+ public:
+  ConvertMemRefAllocOpToDispatchOpPattern(LLVMTypeConverter& type_converter)
+      : ConvertOpToLLVMPattern<memref::AllocOp>(type_converter) {}
+
+ private:
+  // TODO(disc): Remove strides computation.
+  MemRefDescriptor CreateMemRefDescriptor(Location loc,
+                                          ConversionPatternRewriter& rewriter,
+                                          MemRefType memref_type,
+                                          Value allocated_byte_ptr,
+                                          ArrayRef<Value> sizes) const {
+    auto memref_desc = MemRefDescriptor::undef(
+        rewriter, loc, typeConverter->convertType(memref_type));
+
+    // TF AllocateRaw returns aligned pointer => AllocatedPtr == AlignedPtr.
+    Value allocated_type_ptr = rewriter.create<LLVM::BitcastOp>(
+        loc, getElementPtrType(memref_type), allocated_byte_ptr);
+    memref_desc.setAllocatedPtr(rewriter, loc, allocated_type_ptr);
+    memref_desc.setAlignedPtr(rewriter, loc, allocated_type_ptr);
+    memref_desc.setConstantOffset(rewriter, loc, 0);
+
+    if (memref_type.getRank() == 0) {
+      return memref_desc;
+    }
+
+    // Compute strides and populate descriptor `size` and `stride` fields.
+    Value stride_carried = createIndexConstant(rewriter, loc, 1);
+    for (int pos = sizes.size() - 1; pos >= 0; --pos) {
+      Value size = sizes[pos];
+      memref_desc.setSize(rewriter, loc, pos, size);
+      memref_desc.setStride(rewriter, loc, pos, stride_carried);
+      // Update stride
+      if (pos > 0) {
+        stride_carried =
+            rewriter.create<LLVM::MulOp>(loc, stride_carried, size);
+      }
+    }
+    return memref_desc;
+  }
+  LogicalResult matchAndRewrite(
+      memref::AllocOp alloc_op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override;
+};
+
+// Emits LLVM IR to malloc a device memory.
+LogicalResult ConvertMemRefAllocOpToDispatchOpPattern::matchAndRewrite(
+    memref::AllocOp alloc_op, ArrayRef<Value> operands,
+    ConversionPatternRewriter& rewriter) const {
+  mlir::Operation* op = alloc_op.getOperation();
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+
+  Location loc = op->getLoc();
+
+  // check address space
+  auto memref = alloc_op.getResult();
+  MemRefType memref_type = memref.getType().cast<MemRefType>();
+  Attribute memorySpace = memref_type.getMemorySpace();
+  if (!memorySpace) {
+    return failure();
+  }
+  if (!memorySpace.isa<StringAttr>()) {
+    return failure();
+  }
```

Comment:
You're not checking much on the memory space here? It can be any value?

Reply:
Sorry, Mehdi this PR is still WIP. I will refine it later.

Reply:
Done.